File: server/routers/hybridValuation.ts
/**
 * Hybrid Valuation tRPC Router
 * Exposes hybrid valuation model to frontend
 */

import { z } from "zod";
import { publicProcedure, router } from "../_core/trpc";
import { getHybridValuationClient } from "../clients/hybridValuationClient";
import { getDb } from "../db";
import {
  hybridValuations,
  confidenceScores,
  valuationDataSources,
  satelliteImageryAnalysis,
  alternativeDataSources,
} from "../../drizzle/schema";
import { eq, desc } from "drizzle-orm";

// Input validation schemas
const propertyDataSchema = z.object({
  id: z.string().optional(),
  latitude: z.number(),
  longitude: z.number(),
  city: z.string(),
  state: z.string(),
  type: z.string(),
  size_sqm: z.number().optional(),
  bedrooms: z.number().optional(),
  bathrooms: z.number().optional(),
  base_price_per_sqm: z.number().optional(),
});

const valuationRequestSchema = z.object({
  property_data: propertyDataSchema,
  comparable_count: z.number().default(0),
  transaction_history_count: z.number().default(0),
  market_volatility: z.number().default(0.15),
});

export const hybridValuationRouter = router({
  /**
   * Get hybrid valuation for a property
   */
  getValuation: publicProcedure
    .input(valuationRequestSchema)
    .mutation(async ({ input }) => {
      const client = getHybridValuationClient();
      
      try {
        // Call Python hybrid valuation service
        const result = await client.valueProperty(input);
        
        // Store result in database
        const db = await getDb();
        if (db && input.property_data.id) {
          try {
            // Insert hybrid valuation record
            const [hybridVal] = await db.insert(hybridValuations).values({
              valuationId: 0, // Will be linked to base valuation if needed
              propertyId: parseInt(input.property_data.id),
              pathwayUsed: result.pathway_used as "data_rich" | "data_scarce" | "hybrid",
              dataAvailabilityScore: Math.round(result.data_availability_score * 100),
              components: JSON.stringify(result.confidence_details.data_source_contributions),
              finalValuation: Math.round(result.final_valuation),
              confidenceScore: Math.round(result.confidence_score * 100),
              uncertaintyRangeLower: Math.round(result.uncertainty_range[0]),
              uncertaintyRangeUpper: Math.round(result.uncertainty_range[1]),
              satelliteAnalysisId: null,
              alternativeDataId: null,
              modelVersion: result.model_version,
              metadata: JSON.stringify({
                num_components: result.num_components,
                component_methods: result.component_methods,
                valuation_timestamp: result.valuation_timestamp,
              }),
            });
            
            // Insert confidence scores
            if (hybridVal.insertId) {
              await db.insert(confidenceScores).values({
                valuationId: hybridVal.insertId,
                overallConfidence: Math.round(result.confidence_details.overall_confidence * 100),
                confidenceLevel: result.confidence_details.confidence_level as any,
                dataCompletenessScore: Math.round(result.confidence_details.data_completeness.overall_completeness * 100),
                modelAccuracyScore: Math.round(result.confidence_details.confidence_breakdown.model_accuracy_contribution * 100),
                comparableQualityScore: Math.round(result.confidence_details.confidence_breakdown.comparable_quality_contribution * 100),
                satelliteConfidenceScore: result.satellite_analysis 
                  ? Math.round(result.satellite_analysis.confidence_score * 100)
                  : null,
                marketStabilityScore: Math.round(result.confidence_details.confidence_breakdown.market_stability_contribution * 100),
                comparableSalesCount: result.confidence_details.data_completeness.comparable_sales_count,
                transactionHistoryCount: result.confidence_details.data_completeness.transaction_history_count,
                satelliteDataAvailable: result.satellite_analysis ? 1 : 0,
                alternativeDataSourcesCount: result.confidence_details.data_completeness.alternative_data_sources,
                predictionIntervalLower: Math.round(result.confidence_details.uncertainty_metrics.prediction_interval_lower),
                predictionIntervalUpper: Math.round(result.confidence_details.uncertainty_metrics.prediction_interval_upper),
                intervalWidthPercent: Math.round(result.confidence_details.uncertainty_metrics.interval_width_percent),
                standardError: Math.round(result.confidence_details.uncertainty_metrics.standard_error),
                qualityFlags: JSON.stringify(result.confidence_details.quality_flags),
                limitingFactors: JSON.stringify(result.confidence_details.confidence_breakdown.limiting_factors),
                recommendations: JSON.stringify(result.confidence_details.recommendations),
              });
              
              // Insert data source contributions
              for (const contrib of result.confidence_details.data_source_contributions) {
                await db.insert(valuationDataSources).values({
                  valuationId: hybridVal.insertId,
                  sourceName: contrib.source_name,
                  sourceType: contrib.source_name.includes('satellite') ? 'satellite_imagery' :
                             contrib.source_name.includes('comparable') ? 'comparable_sales' :
                             contrib.source_name.includes('market') ? 'market_listings' :
                             contrib.source_name.includes('hedonic') ? 'hedonic_model' : 'comparable_sales',
                  weight: Math.round(contrib.weight * 100),
                  confidence: Math.round(contrib.confidence * 100),
                  valueContribution: Math.round(contrib.value_contribution),
                  dataQuality: result.confidence_details.data_completeness.quality_flag as any,
                  dataCount: null,
                  metadata: JSON.stringify({ importance_rank: contrib.importance_rank }),
                });
              }
              
              // Insert satellite analysis if available
              if (result.satellite_analysis) {
                await db.insert(satelliteImageryAnalysis).values({
                  propertyId: parseInt(input.property_data.id),
                  buildingFootprintSqm: Math.round(result.satellite_analysis.building_footprint_sqm),
                  estimatedHeightM: Math.round(result.satellite_analysis.estimated_height_m * 100), // cm
                  numFloors: result.satellite_analysis.num_floors,
                  roofMaterial: result.satellite_analysis.roof_material as any,
                  roofCondition: result.satellite_analysis.roof_condition as any,
                  buildingDensity: Math.round(result.satellite_analysis.building_density * 10),
                  greenSpaceRatio: Math.round(result.satellite_analysis.green_space_ratio * 100),
                  roadAccessQuality: result.satellite_analysis.road_access_quality as any,
                  amenityDensity: null,
                  roadDensityKm: null,
                  commercialRatio: null,
                  infrastructureScore: null,
                  analysisConfidence: Math.round(result.satellite_analysis.confidence_score * 100),
                  dataQuality: result.confidence_details.data_completeness.quality_flag as any,
                  dataSources: JSON.stringify(['satellite_imagery']),
                  analysisTimestamp: new Date(),
                  valuationMultiplier: null,
                  metadata: JSON.stringify({}),
                });
              }
            }
          } catch (dbError) {
            console.error('Failed to store valuation in database:', dbError);
            // Continue even if DB storage fails
          }
        }
        
        return result;
      } catch (error) {
        console.error('Hybrid valuation error:', error);
        throw new Error(`Valuation failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
      }
    }),

  /**
   * Get valuation history for a property
   */
  getValuationHistory: publicProcedure
    .input(z.object({
      propertyId: z.number(),
      limit: z.number().default(10),
    }))
    .query(async ({ input }) => {
      const db = await getDb();
      if (!db) {
        throw new Error('Database not available');
      }

      const history = await db
        .select()
        .from(hybridValuations)
        .where(eq(hybridValuations.propertyId, input.propertyId))
        .orderBy(desc(hybridValuations.createdAt))
        .limit(input.limit);

      return history;
    }),

  /**
   * Get confidence score details for a valuation
   */
  getConfidenceDetails: publicProcedure
    .input(z.object({
      valuationId: z.number(),
    }))
    .query(async ({ input }) => {
      const db = await getDb();
      if (!db) {
        throw new Error('Database not available');
      }

      const [confidence] = await db
        .select()
        .from(confidenceScores)
        .where(eq(confidenceScores.valuationId, input.valuationId))
        .limit(1);

      if (!confidence) {
        throw new Error('Confidence scores not found');
      }

      // Get data source contributions
      const sources = await db
        .select()
        .from(valuationDataSources)
        .where(eq(valuationDataSources.valuationId, input.valuationId));

      return {
        ...confidence,
        dataSources: sources,
      };
    }),

  /**
   * Get satellite analysis for a property
   */
  getSatelliteAnalysis: publicProcedure
    .input(z.object({
      propertyId: z.number(),
    }))
    .query(async ({ input }) => {
      const db = await getDb();
      if (!db) {
        throw new Error('Database not available');
      }

      const [analysis] = await db
        .select()
        .from(satelliteImageryAnalysis)
        .where(eq(satelliteImageryAnalysis.propertyId, input.propertyId))
        .orderBy(desc(satelliteImageryAnalysis.createdAt))
        .limit(1);

      return analysis || null;
    }),

  /**
   * Health check for hybrid valuation service
   */
  healthCheck: publicProcedure
    .query(async () => {
      const client = getHybridValuationClient();
      
      try {
        const health = await client.healthCheck();
        return {
          status: 'healthy',
          service: health.service,
          version: health.version,
        };
      } catch (error) {
        return {
          status: 'unhealthy',
          error: error instanceof Error ? error.message : 'Unknown error',
        };
      }
    }),
});

File: services/valuation-service/app/api/hybrid_valuation.py
"""
Hybrid Valuation API Endpoints
Exposes hybrid valuation model via FastAPI
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from typing import Optional, List, Dict
import logging

from app.ml.hybrid_model import get_hybrid_model, ModelPathway
from app.ml.confidence_scorer import get_confidence_scorer

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/hybrid-valuation", tags=["hybrid-valuation"])


# Request/Response Models

class PropertyDataRequest(BaseModel):
    """Property data for valuation"""
    id: Optional[str] = None
    latitude: float = Field(..., description="Property latitude")
    longitude: float = Field(..., description="Property longitude")
    city: str = Field(..., description="City name")
    state: str = Field(..., description="State name")
    type: str = Field(..., description="Property type (house, apartment, etc.)")
    size_sqm: Optional[float] = Field(None, description="Property size in square meters")
    bedrooms: Optional[int] = Field(None, description="Number of bedrooms")
    bathrooms: Optional[int] = Field(None, description="Number of bathrooms")
    base_price_per_sqm: Optional[float] = Field(200000, description="Base price per sqm in NGN")


class ValuationRequest(BaseModel):
    """Request for hybrid valuation"""
    property_data: PropertyDataRequest
    comparable_count: int = Field(0, description="Number of comparable properties available")
    transaction_history_count: int = Field(0, description="Number of historical transactions")
    market_volatility: float = Field(0.15, description="Market volatility (0-1)")


class DataSourceContributionResponse(BaseModel):
    """Data source contribution details"""
    source_name: str
    weight: float
    confidence: float
    value_contribution: float
    importance_rank: int


class ConfidenceBreakdownResponse(BaseModel):
    """Confidence breakdown details"""
    data_completeness_contribution: float
    model_accuracy_contribution: float
    comparable_quality_contribution: float
    satellite_confidence_contribution: float
    market_stability_contribution: float
    overall_confidence: float
    confidence_level: str
    limiting_factors: List[str]


class UncertaintyMetricsResponse(BaseModel):
    """Uncertainty metrics"""
    prediction_interval_lower: float
    prediction_interval_upper: float
    interval_width_percent: float
    standard_error: float
    coefficient_of_variation: float
    uncertainty_sources: Dict[str, float]


class DataCompletenessResponse(BaseModel):
    """Data completeness metrics"""
    comparable_sales_count: int
    comparable_sales_score: float
    transaction_history_count: int
    transaction_history_score: float
    satellite_data_available: bool
    satellite_data_score: float
    alternative_data_sources: int
    alternative_data_score: float
    overall_completeness: float
    quality_flag: str


class ConfidenceScoreResponse(BaseModel):
    """Complete confidence score result"""
    overall_confidence: float
    confidence_level: str
    data_completeness: DataCompletenessResponse
    confidence_breakdown: ConfidenceBreakdownResponse
    uncertainty_metrics: UncertaintyMetricsResponse
    data_source_contributions: List[DataSourceContributionResponse]
    quality_flags: List[str]
    recommendations: List[str]


class SatelliteAnalysisResponse(BaseModel):
    """Satellite analysis summary"""
    building_footprint_sqm: float
    estimated_height_m: float
    num_floors: int
    roof_material: str
    roof_condition: str
    building_density: float
    green_space_ratio: float
    road_access_quality: str
    confidence_score: float


class AlternativeDataResponse(BaseModel):
    """Alternative data summary"""
    data_completeness_score: float
    sources_used: List[str]
    has_economic_indicators: bool
    has_market_listing_data: bool
    has_neighborhood_quality: bool


class HybridValuationResponse(BaseModel):
    """Complete hybrid valuation response"""
    final_valuation: float
    confidence_score: float
    uncertainty_range: tuple[float, float]
    pathway_used: str
    data_availability_score: float
    
    # Detailed metrics
    confidence_details: ConfidenceScoreResponse
    satellite_analysis: Optional[SatelliteAnalysisResponse]
    alternative_data: Optional[AlternativeDataResponse]
    
    # Metadata
    model_version: str
    valuation_timestamp: str
    num_components: int
    component_methods: List[str]


# API Endpoints

@router.post("/value", response_model=HybridValuationResponse)
async def value_property(request: ValuationRequest):
    """
    Value a property using the hybrid valuation model
    
    This endpoint automatically selects the best valuation pathway based on
    data availability:
    - Data-Rich: Traditional ML with sufficient comparables
    - Data-Scarce: Satellite imagery + alternative data
    - Hybrid: Ensemble of both approaches
    """
    try:
        logger.info(f"Hybrid valuation request for property in {request.property_data.city}, {request.property_data.state}")
        
        # Get hybrid model
        hybrid_model = get_hybrid_model()
        
        # Convert request to property data dict
        property_data = request.property_data.model_dump()
        
        # Perform hybrid valuation
        valuation_result = await hybrid_model.value_property(
            property_data=property_data,
            comparable_count=request.comparable_count,
            transaction_history_count=request.transaction_history_count
        )
        
        # Calculate detailed confidence scores
        confidence_scorer = get_confidence_scorer()
        confidence_result = confidence_scorer.calculate_confidence(
            valuation_result=valuation_result,
            comparable_count=request.comparable_count,
            transaction_count=request.transaction_history_count,
            market_volatility=request.market_volatility
        )
        
        # Build response
        response = HybridValuationResponse(
            final_valuation=valuation_result.final_valuation,
            confidence_score=valuation_result.confidence_score,
            uncertainty_range=valuation_result.uncertainty_range,
            pathway_used=valuation_result.pathway_used.value,
            data_availability_score=valuation_result.data_availability.overall_score,
            
            # Confidence details
            confidence_details=ConfidenceScoreResponse(
                overall_confidence=confidence_result.overall_confidence,
                confidence_level=confidence_result.confidence_level.value,
                data_completeness=DataCompletenessResponse(
                    comparable_sales_count=confidence_result.data_completeness.comparable_sales_count,
                    comparable_sales_score=confidence_result.data_completeness.comparable_sales_score,
                    transaction_history_count=confidence_result.data_completeness.transaction_history_count,
                    transaction_history_score=confidence_result.data_completeness.transaction_history_score,
                    satellite_data_available=confidence_result.data_completeness.satellite_data_available,
                    satellite_data_score=confidence_result.data_completeness.satellite_data_score,
                    alternative_data_sources=confidence_result.data_completeness.alternative_data_sources,
                    alternative_data_score=confidence_result.data_completeness.alternative_data_score,
                    overall_completeness=confidence_result.data_completeness.overall_completeness,
                    quality_flag=confidence_result.data_completeness.quality_flag.value
                ),
                confidence_breakdown=ConfidenceBreakdownResponse(
                    data_completeness_contribution=confidence_result.confidence_breakdown.data_completeness_contribution,
                    model_accuracy_contribution=confidence_result.confidence_breakdown.model_accuracy_contribution,
                    comparable_quality_contribution=confidence_result.confidence_breakdown.comparable_quality_contribution,
                    satellite_confidence_contribution=confidence_result.confidence_breakdown.satellite_confidence_contribution,
                    market_stability_contribution=confidence_result.confidence_breakdown.market_stability_contribution,
                    overall_confidence=confidence_result.confidence_breakdown.overall_confidence,
                    confidence_level=confidence_result.confidence_breakdown.confidence_level.value,
                    limiting_factors=confidence_result.confidence_breakdown.limiting_factors
                ),
                uncertainty_metrics=UncertaintyMetricsResponse(
                    prediction_interval_lower=confidence_result.uncertainty_metrics.prediction_interval_lower,
                    prediction_interval_upper=confidence_result.uncertainty_metrics.prediction_interval_upper,
                    interval_width_percent=confidence_result.uncertainty_metrics.interval_width_percent,
                    standard_error=confidence_result.uncertainty_metrics.standard_error,
                    coefficient_of_variation=confidence_result.uncertainty_metrics.coefficient_of_variation,
                    uncertainty_sources=confidence_result.uncertainty_metrics.uncertainty_sources
                ),
                data_source_contributions=[
                    DataSourceContributionResponse(
                        source_name=contrib.source_name,
                        weight=contrib.weight,
                        confidence=contrib.confidence,
                        value_contribution=contrib.value_contribution,
                        importance_rank=contrib.importance_rank
                    )
                    for contrib in confidence_result.data_source_contributions
                ],
                quality_flags=confidence_result.quality_flags,
                recommendations=confidence_result.recommendations
            ),
            
            # Satellite analysis summary
            satellite_analysis=SatelliteAnalysisResponse(
                building_footprint_sqm=valuation_result.satellite_analysis.building_features.building_footprint_sqm,
                estimated_height_m=valuation_result.satellite_analysis.building_features.estimated_height_m,
                num_floors=valuation_result.satellite_analysis.building_features.num_floors,
                roof_material=valuation_result.satellite_analysis.building_features.roof_material,
                roof_condition=valuation_result.satellite_analysis.building_features.roof_condition,
                building_density=valuation_result.satellite_analysis.building_features.building_density,
                green_space_ratio=valuation_result.satellite_analysis.building_features.green_space_ratio,
                road_access_quality=valuation_result.satellite_analysis.building_features.road_access_quality,
                confidence_score=valuation_result.satellite_analysis.building_features.confidence_score
            ) if valuation_result.satellite_analysis else None,
            
            # Alternative data summary
            alternative_data=AlternativeDataResponse(
                data_completeness_score=valuation_result.alternative_data.data_completeness_score,
                sources_used=valuation_result.alternative_data.sources_used,
                has_economic_indicators=valuation_result.alternative_data.economic_indicators is not None,
                has_market_listing_data=valuation_result.alternative_data.market_listing_data is not None,
                has_neighborhood_quality=valuation_result.alternative_data.neighborhood_quality is not None
            ) if valuation_result.alternative_data else None,
            
            # Metadata
            model_version=valuation_result.metadata.get('model_version', '2.0-hybrid'),
            valuation_timestamp=valuation_result.metadata.get('valuation_date', ''),
            num_components=valuation_result.metadata.get('num_components', 0),
            component_methods=valuation_result.metadata.get('component_methods', [])
        )
        
        logger.info(f"Hybrid valuation completed: NGN {response.final_valuation:,.0f} "
                   f"(confidence: {response.confidence_score:.2%}, pathway: {response.pathway_used})")
        
        return response
        
    except Exception as e:
        logger.error(f"Hybrid valuation failed: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Valuation failed: {str(e)}")


@router.get("/health")
async def health_check():
    """Health check endpoint"""
    return {
        "status": "healthy",
        "service": "hybrid-valuation",
        "version": "2.0"
    }
